src/rgw/rgw_dedup.cc

```diff
@@ -0,0 +1,2496 @@
+      ldpp_dout(dpp, 1) << __func__ << "::ERROR: corrupted etag" << dendl;
+      return -EINVAL;
+    }
+
+    if (unlikely((cct->_conf->subsys.should_gather<ceph_subsys_rgw_dedup, 20>()))) {
+      show_ingress_bucket_idx_obj(dpp, parsed_etag, p_bucket->get_name(), entry.key.name);
+    }
+
+    // TBD: what should be done about storage_class and how do we check for this ??
+    // Should we accept *any* storage class while limiting dedup with storage_class?
+    // Should we use a skip-list of storage_classes we should skip (like glacier) ?
+    const std::string& storage_class = rgw_placement_rule::get_canonical_storage_class(entry.meta.storage_class);
+    if (storage_class == RGW_STORAGE_CLASS_STANDARD) {
+      p_worker_stats->default_storage_class_objs++;
+      p_worker_stats->default_storage_class_objs_bytes += ondisk_byte_size;
+    }
+    else {
+      ldpp_dout(dpp, 10) << __func__ << "::" << entry.key.name
+                         << "::storage_class:" << entry.meta.storage_class << dendl;
+      p_worker_stats->non_default_storage_class_objs++;
+      p_worker_stats->non_default_storage_class_objs_bytes += ondisk_byte_size;
+      return 0;
+    }
+
+    if (ondisk_byte_size <= d_min_obj_size_for_dedup) {
+      if (parsed_etag.num_parts == 0) {
+        // dedup only useful for objects bigger than 4MB
+        p_worker_stats->ingress_skip_too_small++;
+        p_worker_stats->ingress_skip_too_small_bytes += ondisk_byte_size;
+
+        if (ondisk_byte_size >= 64*1024) {
+          p_worker_stats->ingress_skip_too_small_64KB++;
+          p_worker_stats->ingress_skip_too_small_64KB_bytes += ondisk_byte_size;
+        }
+        return 0;
+      }
+      else {
+        // multipart objects are always good candidates for dedup
+        // the head object is empty and data is stored only in tail objs
+        p_worker_stats->small_multipart_obj++;
+      }
+    }
+    // multipart/single_part counters are for objects being fully processed
+    if (parsed_etag.num_parts > 0) {
+      p_worker_stats->multipart_objs++;
+    }
+    else {
+      p_worker_stats->single_part_objs++;
+    }
+
+    return add_disk_rec_from_bucket_idx(disk_arr, p_bucket, &parsed_etag,
+                                        entry.key.name, entry.meta.size);
+  }
+
+  //---------------------------------------------------------------------------
+  void Background::check_and_update_heartbeat(unsigned shard_id, uint64_t count_a,
+                                              uint64_t count_b, const char *prefix)
+  {
+    utime_t now = ceph_clock_now();
+    utime_t time_elapsed = now - d_heart_beat_last_update;
+    if (unlikely(time_elapsed.tv.tv_sec >= d_heart_beat_max_elapsed_sec)) {
+      ldpp_dout(dpp, 20) << __func__ << "::max_elapsed_sec="
+                         << d_heart_beat_max_elapsed_sec << dendl;
+      d_heart_beat_last_update = now;
+      d_cluster.update_shard_token_heartbeat(d_dedup_cluster_ioctx, shard_id,
```

Comment:
why do you ignore failure indication returned from this call?

Reply:
Failing to update heartbeat doesn't break anything as long as we can keep processing the data (if not we will bailout).
Heartbeats are updated every 3 seconds and we can withstand multiple failures .

Even if we fail for long time there will be no race as other RGW are not allowed to take over tokens and in worst case we will lose dedup opportunities